File: articles/water-energy-material-conditions-of-life/python/homeostatic_setpoint.py
"""
Homeostatic setpoint simulation workflow.

Run:
    python python/homeostatic_setpoint.py
"""

from __future__ import annotations

from pathlib import Path

import numpy as np
import pandas as pd

from water_energy_core import homeostatic_solution


ARTICLE_DIR = Path(__file__).resolve().parents[1]
HOMEOSTASIS_PATH = ARTICLE_DIR / "data" / "homeostasis_scenarios.csv"


def main() -> None:
    """Simulate homeostatic return across scenarios."""

    scenarios = pd.read_csv(HOMEOSTASIS_PATH)

    rows = []

    for _, scenario in scenarios.iterrows():
        time = np.arange(0, scenario["time_end"] + scenario["dt"], scenario["dt"])

        state = homeostatic_solution(
            time,
            scenario["initial_value"],
            scenario["setpoint"],
            scenario["correction_rate"],
        )

        rows.append(
            {
                "scenario": scenario["scenario"],
                "initial_value": scenario["initial_value"],
                "setpoint": scenario["setpoint"],
                "correction_rate": scenario["correction_rate"],
                "final_state": state[-1],
                "final_deviation": state[-1] - scenario["setpoint"],
                "half_recovery_time": (
                    np.log(2) / scenario["correction_rate"]
                    if scenario["correction_rate"] > 0
                    else np.nan
                ),
            }
        )

    print(pd.DataFrame(rows).round(5).to_string(index=False))


if __name__ == "__main__":
    main()

File: articles/water-energy-material-conditions-of-life/r/homeostatic_setpoint.R
# Homeostatic setpoint workflow in R.

scenario_path <- file.path("data", "homeostasis_scenarios.csv")

if (!file.exists(scenario_path)) {
  scenario_path <- file.path("..", "data", "homeostasis_scenarios.csv")
}

scenarios <- read.csv(scenario_path)

simulate_homeostasis <- function(initial_value, setpoint, correction_rate, time_end, dt) {
  time <- seq(0, time_end, by = dt)
  state <- setpoint + (initial_value - setpoint) * exp(-correction_rate * time)

  data.frame(
    time = time,
    state = state,
    setpoint = setpoint,
    deviation = state - setpoint
  )
}

summary_rows <- lapply(seq_len(nrow(scenarios)), function(i) {
  s <- scenarios[i, ]

  sim <- simulate_homeostasis(
    s$initial_value,
    s$setpoint,
    s$correction_rate,
    s$time_end,
    s$dt
  )

  data.frame(
    scenario = s$scenario,
    final_state = tail(sim$state, 1),
    final_deviation = tail(sim$deviation, 1),
    half_recovery_time = ifelse(s$correction_rate > 0, log(2) / s$correction_rate, NA)
  )
})

summary_df <- do.call(rbind, summary_rows)

print(round(summary_df, 5))
